app/helpers/task_list_helper.rb

```diff
@@ -0,0 +1,129 @@
+# frozen_string_literal: true
+
+module TaskListHelper
+  def row_helper(trainee, task)
+    task_list = {
+      publish_course_details:
+        {
+          task_name: "Course details",
+          path: edit_trainee_publish_course_details_path(trainee),
+          confirm_path: trainee_course_details_confirm_path(trainee),
+          classes: "course-details",
+          status: ProgressService.call(
+            validator: ValidatePublishCourseForm.new(trainee),
+            progress_value: trainee.progress.course_details,
+          ).status,
+        },
+
+      course_details:
+        {
+          task_name: "Course details",
+          path: edit_trainee_course_details_path(trainee),
+          confirm_path: trainee_course_details_confirm_path(trainee),
+          classes: "course-details",
+          status: ProgressService.call(
+            validator: CourseDetailsForm.new(trainee),
+            progress_value: trainee.progress.course_details,
+          ).status,
+        },
+
+      training_details:
+        {
+          task_name: "Training details",
+          path: edit_trainee_training_details_path(trainee),
+          confirm_path: trainee_training_details_confirm_path(trainee),
+          classes: "training-details",
+          hint_text: "Start date and trainee ID",
+          status: ProgressService.call(
+            validator: TrainingDetailsForm.new(trainee),
+            progress_value: trainee.progress.training_details,
+          ).status,
+        },
+
+      placement_details:
+        {
+          task_name: "Placement details",
+          path: review_draft_trainee_path(trainee),
+          confirm_path: review_draft_trainee_path(trainee),
+          classes: "placement-details",
+          status: ProgressService.call(
+            validator: PlacementDetailForm.new(trainee),
+            progress_value: trainee.progress.placement_details,
+          ).status,
+        },
+
+      school_details:
+        {
+          task_name: "Lead and employing schools",
+          path: edit_trainee_lead_schools_path(trainee),
+          confirm_path: edit_trainee_lead_schools_path(trainee),
+          classes: "school-details",
+          status: ProgressService.call(
+            validator: Schools::FormValidator.new(trainee),
+            progress_value: trainee.progress.schools,
+          ).status,
+        },
+    }
+
+    unless trainee.apply_application?
+      return if task_list[:training_details][:task_name] != "Training details"
+
+      task_list[:training_details][:task_name] = "Trainee start date and ID"
+    end
+
+    task_list[task]
+  end
+end
+
+def draft_only_row_helper(trainee, task)
```

Comment:
Why does this config need to be in a separate helper method? Can these items not live in the `task_list` in the general `row_helper` above?

Reply:
I've put them all into the row helper method 👍